### PR TITLE
[cpp] Add debugging support for C/C++ programs 

### PR DESCRIPTION
**Pull request by marechal-p**
Using `code-debug`: https://github.com/WebFreak001/code-debug

### CODE REVIEW

**Comment by thegecko** [comment]
@marechal-p This is very similar to the device debug adapter we are working on. This is sufficiently different to warrant it's own adapter, however :)
I'm close to opening a PR, so you can see for yourself.

File: packages/cpp/src/node/debug/cpp-debug-contribution.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+ ********************************************************************************/
+
+const path = require('path');
+const packageJson = require('../../../package.json');
+const debugAdapterDir = packageJson['debugAdapter']['dir'];
+
+import { injectable } from 'inversify';
+import { DebugConfiguration } from '@theia/debug/lib/common/debug-common';
+import { DebugAdapterContribution, DebugAdapterExecutable } from '@theia/debug/lib/node/debug-model';
+import * as Ajv from 'ajv';
```
Comment (thegecko) [inline packages/cpp/src/node/debug/cpp-debug-contribution.ts]: We are using `json-schema-defaults` similar to @jonahgraham , I wonder if there is functional crossover between the packages?

File: packages/cpp/src/node/debug/cpp-debug-contribution.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        '[.]crystal$',
+    ];
+
+    provideDebugConfigurations = [
+        {
+            type: 'gdb',
+            request: 'launch',
+            name: 'Launch Program',
+            target: './bin/executable',
+            cwd: '${workspaceRoot}',
```
Comment (thegecko) [inline packages/cpp/src/node/debug/cpp-debug-contribution.ts]: Does expansion of `${workspaceRoot}` work?

File: packages/debug/package.json [hunk 3]
```diff
@@ -12,17 +12,16 @@
     "@theia/process": "^0.3.14",
```
Comment (thegecko) [inline packages/debug/package.json]: Should changes to `theia/debug` be put in a separate PR?

File: packages/cpp/src/node/debug/cpp-debug-contribution.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    ];
+
+    resolveDebugConfiguration(config: DebugConfiguration): DebugConfiguration {
+
+        switch (config.request) {
+            case 'launch': return this.validateConfiguration(cppDebugConfigurationValidators.Launch, config);
+            case 'attach': return this.validateConfiguration(cppDebugConfigurationValidators.Attach, config);
+        }
+
+        throw new Error(`unkmown request: "${config.request}"`);
```
Comment (thegecko) [inline packages/cpp/src/node/debug/cpp-debug-contribution.ts]: type `unknown`

File: packages/cpp/src/node/debug/cpp-debug-contribution.ts (hunk 2 above)
Comment (marechal-p) [reply packages/cpp/src/node/debug/cpp-debug-contribution.ts]: I currently don't know a way to resolve variables from the backend, so no.

Although it should be a thing.

Should I remove the template-like placeholder for now?

File: packages/debug/package.json (hunk 3 above)
Comment (marechal-p) [reply packages/debug/package.json]: I separated in 3 different commits:
- added a named logger for easier debugging of the debug extension when you make contributions
- used the preferred async/await notation while I was at it
- added the cpp debugger

I didn't think the changes I made to the debug extension were worth a dedicated PR, so I made the two extra commits here. If you really want it out I could do it, me I am fine like this :)

File: packages/cpp/src/node/debug/cpp-debug-contribution.ts (hunk 2 above)
Comment (thegecko) [reply packages/cpp/src/node/debug/cpp-debug-contribution.ts]: I think so. Perhaps create an issue to be able to do that? It would be very useful.

File: packages/debug/package.json (hunk 3 above)
Comment (thegecko) [reply packages/debug/package.json]: I don't mind

File: packages/cpp/src/node/debug/cpp-debug-contribution.ts (hunk 2 above)
Comment (marechal-p) [reply packages/cpp/src/node/debug/cpp-debug-contribution.ts]: Actually after skimming through the code, there seems to be support for config resolution, but let me check if this actually works...

File: packages/cpp/src/node/debug/cpp-debug-contribution.ts (hunk 1 above)
Comment (marechal-p) [reply packages/cpp/src/node/debug/cpp-debug-contribution.ts]: It seems like it: https://github.com/epoberezkin/ajv#assigning-defaults

But `ajv` was already used in `core` and `keymaps` packages, this is why I used it here.

File: packages/cpp/src/node/debug/cpp-debug-contribution.ts (hunk 2 above)
Comment (marechal-p) [reply packages/cpp/src/node/debug/cpp-debug-contribution.ts]: Turns out the Theia equivalent for vscode's `workspaceRoot` is `workspaceFolder`. But it works :)

File: packages/cpp/src/node/debug/cpp-debug-contribution.ts (hunk 1 above)
Comment (thegecko) [reply packages/cpp/src/node/debug/cpp-debug-contribution.ts]: nice, i'll see if I can use this instead :)

File: packages/cpp/src/node/debug/cpp-debug-contribution.ts (hunk 2 above)
Comment (thegecko) [reply packages/cpp/src/node/debug/cpp-debug-contribution.ts]: Just found this too, that's awesome!

**Comment by marechal-p** [comment]
@thegecko thanks for the feedback!
Did you catch anything else?

**Comment by thegecko** [comment]
All looks good, would you mind checking out the cortex debug extension?

https://github.com/theia-ide/theia/pull/3000

**Comment by marechal-p** [comment]
@MarkZ3 :)
I only debugged a `while (true)` loop so far, are you able to use this on more complex projects? am noob.

File: packages/cpp/package.json [hunk 5]
```diff
@@ -50,5 +49,9 @@
   },
   "nyc": {
     "extends": "../../configs/nyc.json"
+  },
+  "debugAdapter": {
+    "downloadUrl": "https://github.com/marechal-p/code-debug/releases/download/v0.22.0/code-debug.tar.gz",
```
[the remark this answers is not in the document]
Comment (marechal-p) [reply packages/cpp/package.json]: The original `code-debug` repository owner didn't publish any sort of release, and wasn't planning to. So I forked the repo, and made a release on my own.

In the end I just followed what was already done in `debug-nodejs`: https://github.com/theia-ide/theia/blob/master/packages/debug-nodejs/package.json#L49.

Any concerns?

[the remark this answers is not in the document]
Comment (MarkZ3) [reply packages/cpp/package.json]: Not really. I would think it's better if things are hosted in non-personal repos (under theia-ide?), for bus-factor concerns and build reproducibility. But I guess anyone is free to make a new fork and make a new release anyway if something happens. This also means you take ownership of this I guess ;)

**Comment by MarkZ3** [comment]
@marechal-p I'll try if I have some time. Probably @simark could try debugging clangd or gdb. I don't have an environment to test gdb or Theia (on Linux) at the moment for a sizeable project. It would be useful to know what is supposed to work and what is known to be missing otherwise I might give you a lot of feedback and create issues that you already know about. Perhaps the Readme would be a good place to state that in order to manage user expectations...

File: packages/cpp/package.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (simark) [reply packages/cpp/package.json]: I agree with Marc-André, it also looks more professional :)

**Comment by simark** [comment]
Yes it's on my todo list to give it a try.  We also need to start thinking about a testsuite.  Since we'll be under the Eclipse unbrella soon, can we steal some bits from the CDT DSF-GDB testsuite?

**Comment by MarkZ3** [comment]
I wanted to try it, but now it looks like it was reverted??

**Comment by marcdumais-work** [comment]
> I wanted to try it, but now it looks like it was reverted??

Correct, but for reasons unrelated directly to the feature itself. Basically it made @theia/cpp depend on @theia/debug, which is still early/experimental. So it became impossible to make a Theia app that includes cpp but excludes debug. 

We momentarily reverted the original commit, and will bring the feature back in its own ```cpp-debug``` extension, that can be omitted in Theia apps, if preferred. See https://github.com/theia-ide/theia/pull/3024

File: packages/cpp/package.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (marcdumais-work) [reply packages/cpp/package.json]: Agreed, smt to add to the to-do list. We do have the option to fork ```code-debug``` under ```theia-ide```.

**Comment by marcdumais-work** [comment]
> Since we'll be under the Eclipse unbrella soon, can we steal some bits from the CDT DSF-GDB testsuite?

Might be. Not sure if the license differences (EPLv1 vs EPLv2 + GPL classpath exception) might be an issue.

**Comment by marcdumais-work** [comment]
Note: This PR was then reverted because it caused a dependency issue when used without its implicit dependency ```@theia/debug```.  

So instead of re-submitting here as a new extension, we decided to do so in the new repo created for ```cpp``` under the Eclipse org (the current cpp extension will later be moved to that same repo as well ): 

https://github.com/eclipse/theia-cpp-extension/pull/1
